### PR TITLE
Transform to lower packs/unpacks without transpose, except when the packing occurs on a constant

### PR DESCRIPTION
**Pull request by rolfmorel**
A transform that enables selectively "reverting" packing on `linalg.generic` operands and unpacking on the generic's results by lowering `tensor.pack` to `tensor.expand_shape` with identity permutation and `tensor.unpack` to `tensor.collapse_shape` with identity permutation.

The new pass `-lower-packs-unpacks-without-transpose` does this reverting on all (compatible) packed/unpacked `linalg.generic` operands/results, except if they are constants. Similarly, the CLI flag `--lower-pack-unpack-without-transpose` enables this pass in the default pipeline.

### CODE REVIEW

File: lib/TPP/Transforms/ConvertPackToExpandShape.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: I don't think we can update other users. It might be best to only update specific generic's operand.
Or maybe it doesn't make sense to rewrite pack at all if there are multiple users. But then if these are multiple generic contractions then maybe it's still fine.

For us, restriction to only one user might be enough, in general I'd leave that for a user to decide.

File: lib/TPP/Transforms/ConvertPackToExpandShape.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/ConvertPackToExpandShape.cpp:55]: nit: maybe use `ReassociationIndices` alias

File: lib/TPP/Transforms/ConvertPackToExpandShape.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/ConvertPackToExpandShape.cpp:135]: I'd move it to a separate pattern

Comment (rolfmorel) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp:135]: On the one hand, yes, it might be nice to have it available as a separate pattern. On the other hand, when we modify the out parameter of the linalg, the result type changes/needs to be changed as well. This in turn means the unpack afterwards needs to have _something_ done to it. In the context of a single pattern it is clear what this _something_ is: rewrite it to a collapse_shape. 

In case of separate patterns, do you have an idea what to do with the result of the linalg when we change the shape of the out parameter? If we don't do anything the IR is most likely left in an illegal state (i.e. types will not match up).

Comment (rolfmorel) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp:135]: When it is not the case that there is a single unpack user of the linalg, we probably shouldn't do the rewrite of the out parameter any way (as we don't know how the users are impacted by the shape change of the type). The only sensible thing seems to be to insert some op in between the linalg's result and all the users to invert the shape change.

File: lib/TPP/Transforms/ConvertPackToExpandShape.cpp
Comment (rolfmorel) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: Have now added requirement for single user for there to be a match.

We can choose to relax later.

File: lib/TPP/Transforms/ConvertPackToExpandShape.cpp
Comment (adam-smnk) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: For outs, we definitely need the whole chain pack->generic->unpack. If there's any user other than unpack then it probably doesn't make sense to change outs shape.
Also this case might require extra checks on the pack-unpack pair to see if they are compatible.

So my thinking is mainly that it'd be cleaner to move this extra logic to a separate pattern that focuses on outs, and this one focuses on ins. But maybe it doesn't make much difference.

Comment (rolfmorel) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: Done.

Comment (rolfmorel) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: If you could give an example of a compatibility issue for pack-unpack pairs, that would much appreciated. I had a think and came to the conclusion that as the generic's result type is shared by the pair, this is enough to force compatability.

Comment (adam-smnk) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: Not sure if you can fold a pair in which pack has different perms than unpack `outer_dims_perm = [1, 0] inner_dims_pos = [0, 1]`.

Comment (adam-smnk) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: There are also probably cases when you can fold pack but need to keep unpack. For example, maybe pack can have permutation which then is performed by the generic but unpack has to stay as you need extra data movement anyway.

But let's ignore these for now.

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (adam-smnk) [inline test/Passes/lower-packs-and-unpacks-without-transpose.mlir:42]: I think this constant should have a pack op. Prepacked one doesn't test anything.

**Comment by rolfmorel** [comment]
@adam-smnk , @rengolin 

This is now ready for final review.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:80]: That's an oddly specific pattern. I think the pattern should be either more precise or more generic.

Comment (rolfmorel) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:80]: This is based on your suggestion to have separate patterns.

Comment (adam-smnk) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:80]: I mean the explicit operands 0 and 1

Comment (rolfmorel) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:80]: I see - the current thing is incorrect in case of non-binary input generics. Before I had the guard that this only applied to ContractionInterface ops but I removed it as I thought it was not necessary.

Will fix this so that it iterates over all inputs of the generic (and correspondingly fix the output/unpack pattern).

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (rolfmorel) [reply test/Passes/lower-packs-and-unpacks-without-transpose.mlir:42]: The pre-transform IR for this test is just `-pack-matmul` applied to a matmul where one of the arguments is a constant. The difference between this test and the test after it is exactly the difference between the pack having already been applied or not.

If this test is not useful (as has more to do with the integration between `-pack-matmul` and the new pass, let me know.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:132]: This might not be safe to do if unpack has different permutations than pack.

Comment (adam-smnk) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:132]: Or to put it more simply if you have sth like:
```
%packed = pack(<128x64>)
generic outs(%packed)
unpack into <64x128>
```

File: include/TPP/PassBundles.td
Comment (rolfmorel) [reply include/TPP/PassBundles.td]: Done.

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (rolfmorel) [reply test/Passes/lower-packs-and-unpacks-without-transpose.mlir]: Converted it into actually running `-pack-matmul` on a matmul operating on a constant . This has a bit of integration test flavour, but not on the same scale as most tests in `test/Integration`.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (rolfmorel) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp]: Made both patterns agnostic to the arity. If you could check if this looks right, that would be appreciated.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (rolfmorel) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp]: Have made it such that the pack on the out param and the unpack on the corresponding result must specify the same permutation. Else the pattern does not apply.

File: lib/TPP/Transforms/ConvertPackToExpandShape.cpp
Comment (rolfmorel) [reply lib/TPP/Transforms/ConvertPackToExpandShape.cpp]: Moved to two separate patterns. There's discussion below as to the safe permutations on unpack.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:128]: nit: in case of identity `getOuterDimsPerm` can be empty so it's not completely reliable

**Review by adam-smnk — APPROVED** [review]
Logic looks solid, I'd still cleanup the tests a bit.

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (adam-smnk) [reply test/Passes/lower-packs-and-unpacks-without-transpose.mlir]: That makes tests more unpredictable as unrelated tests start breaking when one of the passes changes due to coupling.

IMO, it's nice to have one (perhaps cleaned up) matmul test case as that's what we care about, otherwise you can verify edge cases on much simpler examples.

**Comment by rolfmorel** [comment]
Performance on the lora-fragment benchmark in use for Openvino:

```
$ python -m xonsh lora-runner.xsh 
CONFIGS [[8], [16], [32], [64], [128], [256], [512], [1024], [2048], [4096], [8192]]
OV NO-MLIR [0.84, 1.38, 2.48, 4.7, 9.62, 19.14, 37.54, 74.51, 147.15, 311.04, 628.24]
OV MLIR [3.82, 4.43, 5.35, 8.06, 13.45, 23.87, 44.33, 85.15, 167.73, 340.55, 690.18]
NO-OV MLIR [4.05, 4.54, 5.57, 8.2, 13.47, 23.31, 44.07, 83.64, 166.25, 333.87, 662.45]
MANUAL MLIR [0.54, 1.1, 2.2, 4.43, 8.82, 17.39, 35.26, 70.28, 138.55, 279.32, 561.73]
```

- OV NO-MLIR vs. MANUAL MLIR compares plain OV vs TPP-MLIR with this new rewrite.
- NO-OV MLIR vs. MANUAL MLIR compares TPP-MLIR vs TPP-MLIR with this new rewrite.
- OV MLIR vs. NO-OV MLIR compares TPP-MLIR vs TPP-MLIR (both without the new rewrite) to infer overhead of calling MLIR-generated code from OV.

**Comment by adam-smnk** [comment]
> NO-OV MLIR vs. MANUAL MLIR compares TPP-MLIR vs TPP-MLIR with this new rewrite.

Kind of, NO-OV MLIR still has runtime packing due to weights passed as arguments.

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (rolfmorel) [reply test/Passes/lower-packs-and-unpacks-without-transpose.mlir]: Only the first test case uses `-pack-matmul` when run - the other tests use IR obtained by manually running `-pack-matmul`. It is also as simple as we can go to test a matmul with a constant and covers the situation that we care about. Hence I believe the tests already are as you suggest.

I think in this case it is also not a bad idea to have such a "coupled" test - though possibly we need to move it to  `test/Integration` - as we would like to know when changes to either pass results in this part of the pipeline no longer doing what we expect.

**Comment by rolfmorel** [comment]
> `> NO-OV MLIR vs. MANUAL MLIR compares TPP-MLIR vs TPP-MLIR with this new rewrite.
> 
> Kind of, NO-OV MLIR still has runtime packing due to weights passed as arguments.

Disambiguation issue: NO-OV MLIR vs. MANUAL MLIR compares (TPP-MLIR) vs (TPP-MLIR with this new rewrite).

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (adam-smnk) [reply test/Passes/lower-packs-and-unpacks-without-transpose.mlir]: > Only the first test case uses -pack-matmul when run - the other tests use IR obtained by manually running -pack-matmul

Then the one obtained from a manual run should be enough. Otherwise, the pass can be tested with more readable 1 input 1 output cases + maybe sth for multiple results if you wanna stress that.

> I think in this case it is also not a bad idea to have such a "coupled" test - though possibly we need to move it to test/Integration

It's one of those things that are good in theory but super annoying in practice. And that's what we've started with as seen by the current collection of integration and bundle tests. In reality, they end up being brittle and whenever sth breaks I try to at least simplify or tweak them. Although, most could be removed with no loss.

I'd be all for integration tests that validate full execution by examining actual computed non-trivial results to double check correctness. I think it's sth that we lack currently.

> as we would like to know when changes to either pass results in this part of the pipeline no longer doing what we expect.

I think it'd be more helpful to verify behavior of the whole pipeline on more known kernels.
Overall, that is good goal but I don't think manually made "test pipelines" truly achieve it.

Comment (adam-smnk) [reply test/Passes/lower-packs-and-unpacks-without-transpose.mlir]: But all in all, it's subjective. I'm not gonna block on that - if you prefer to have such multi pass example, I'm fine with it in integration tests.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (rolfmorel) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:128]: Fixed.

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (rolfmorel) [reply test/Passes/lower-packs-and-unpacks-without-transpose.mlir]: Made an integration test to test functional correctness in case of a matmul with a constant as argument. Replaced the `-pack-matmul` test case with a single argument linalg.generic test.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (rolfmorel) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp]: A sanity check on this would be appreciated - thanks!

**Comment by rolfmorel** [comment]
@adam-smnk, @rengolin, in case of no further comments I will merge this tomorrow, Sept 25th.

File: test/Integration/lower-pack-unpack-folding-transpose.mlir
Comment (adam-smnk) [inline test/Integration/lower-pack-unpack-folding-transpose.mlir:1]: nit: a bit outdated output name ;)
Maybe sth more specific to the test name?

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (adam-smnk) [inline test/Passes/lower-packs-and-unpacks-without-transpose.mlir:1]: Should this line be removed?

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:135]:
```suggestion
        std::iota(packOuterDimsPerm.begin(), packOuterDimsPerm.end(), 0);
```

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (adam-smnk) [inline lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:141]:
```suggestion
        std::iota(unpackOuterDimsPerm.begin(), unpackOuterDimsPerm.end(), 0);
```

Comment (adam-smnk) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp:141]: Could you add a test case for this?

File: test/Passes/lower-packs-and-unpacks-without-transpose.mlir
Comment (adam-smnk) [inline test/Passes/lower-packs-and-unpacks-without-transpose.mlir:1]: You should verify affine maps in at least one case to check that transpose is folded correctly.

File: lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp
Comment (adam-smnk) [reply lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp]: Approach is fine, just a small fix in the logic is needed.

**Comment by rolfmorel** [comment]
Thank you for the thorough review, @adam-smnk !

I think I have addressed everything, though will wait a couple more hours before I merge.

File: test/Integration/lower-pack-unpack-without-transpose.mlir
Comment (adam-smnk) [inline test/Integration/lower-pack-unpack-without-transpose.mlir:6]: That's way too large for CI, single run alone on my local machine takes 25% time of whole test suite.

**Review by adam-smnk — COMMENTED** [review]
Thanks for the tweaks 👍
Looks good

**Pull request merged**
